framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java

```diff
@@ -58,38 +73,70 @@
      *     argument named "value" whose type is a String array.
      * @param bottom the bottom type in the hierarchy, which must be a subtype of {@code
      *     accumulator}. The bottom type should be an annotation with no arguments.
+     * @param predicate the predicate annotation. Either null (if predicates are not supported), or
+     *     an annotation with a single element named "value" whose type is a String.
      */
     protected AccumulationAnnotatedTypeFactory(
             BaseTypeChecker checker,
             Class<? extends Annotation> accumulator,
-            Class<? extends Annotation> bottom) {
+            Class<? extends Annotation> bottom,
+            @Nullable Class<? extends Annotation> predicate) {
         super(checker);
 
         this.accumulator = accumulator;
-
         // Check that the requirements of the accumulator are met.
         Method[] accDeclaredMethods = accumulator.getDeclaredMethods();
         if (accDeclaredMethods.length != 1) {
             rejectMalformedAccumulator("have exactly one element");
         }
-        Method value = accDeclaredMethods[0];
-        if (value.getName() != "value") { // interned
+
+        Method accValue = accDeclaredMethods[0];
+        if (accValue.getName() != "value") { // interned
             rejectMalformedAccumulator("name its element \"value\"");
         }
-        if (!value.getReturnType().isInstance(new String[0])) {
+        if (!accValue.getReturnType().isInstance(new String[0])) {
             rejectMalformedAccumulator("have an element of type String[]");
         }
-        if (((String[]) value.getDefaultValue()).length != 0) {
+        if (((String[]) accValue.getDefaultValue()).length != 0) {
             rejectMalformedAccumulator("have the empty String array {} as its default value");
         }
 
+        this.predicate = predicate;
+        // If there is a predicate annotation, check that its requirements are met.
+        if (predicate != null) {
+            Method[] predDeclaredMethods = predicate.getDeclaredMethods();
+            if (predDeclaredMethods.length != 1) {
+                rejectMalformedPredicate("have exactly one element");
+            }
+            Method predValue = predDeclaredMethods[0];
+            if (predValue.getName() != "value") { // interned
+                rejectMalformedPredicate("name its element \"value\"");
+            }
+            if (!predValue.getReturnType().isInstance("")) {
+                rejectMalformedPredicate("have an element of type String");
+            }
+        }
+
         this.bottom = AnnotationBuilder.fromClass(elements, bottom);
         this.top = createAccumulatorAnnotation(Collections.emptyList());
 
-        // Every subclass must call postInit!  This does not do so for subclasses.
-        if (this.getClass() == AccumulationAnnotatedTypeFactory.class) {
-            this.postInit();
```

Comment:
Wall calling `postInit()` here before a bug?

Reply:
The condition was always false, because this is an abstract class. All subclasses have to call `postInit`.